Mathlib/Analysis/Meromorphic/Divisor/Basic.lean

```diff
@@ -0,0 +1,424 @@
+/-
+Copyright (c) 2025 Stefan Kebekus. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Stefan Kebekus
+-/
+import Mathlib.Analysis.Normed.Field.Basic
+import Mathlib.Analysis.SpecialFunctions.Log.Basic
+
+/-!
+# Divisors on subsets of normed fields
+
+This file defines divisors, a standard book-keeping tool in complex analysis
+used to keep track of pole/vanishing orders of meromorphic objects, typically
+functions or differential forms. It provides supporting API and establishes
+divisors as an instance of a lattice ordered commutative group.
+
+Throughout the present file, `𝕜` denotes a nontrivially normed field, and `U` a
+subset of `𝕜`.
+
+## TODOs
+
+- Constructions: The divisor of a meromorphic function, behavior under product
+  of meromorphic functions, behavior under addition, behavior under restriction
+- Construction: The divisor of a rational polynomial
+-/
+
+open Filter Metric Real Set Topology
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜] {U : Set 𝕜}
+
+/-!
+## Definition, coercion to functions and basic extensionality lemmas
+
+A divisor on `U` is a function `𝕜 → ℤ` whose support is discrete within `U` and
+entirely contained within `U`.  The theorem
+`supportDiscreteWithin_iff_locallyFiniteWithin` shows that this is equivalent to
+the textbook definition, which requires the support of `f` to be locally finite
+within `U`.
+-/
+
+/-- A divisor on `U` is a triple specified below. -/
+structure Divisor (U : Set 𝕜) where
+  /-- A function `𝕜 → ℤ` -/
+  toFun : 𝕜 → ℤ
+  /-- A proof that the support of `toFun` is contained in `U` -/
+  supportWithinDomain : toFun.support ⊆ U
+  /-- A proof the the support is discrete within `U` -/
+  supportDiscreteWithinDomain : toFun =ᶠ[codiscreteWithin U] 0
+
+/-- The condition `supportDiscreteWithinU` in a divisor is equivalent to saying
+that the support is locally finite near every point of `U`. -/
+theorem supportDiscreteWithin_iff_locallyFiniteWithin {f : 𝕜 → ℤ} (h : f.support ⊆ U) :
+    f =ᶠ[codiscreteWithin U] 0 ↔ ∀ z ∈ U, ∃ t ∈ 𝓝 z, Set.Finite (t ∩ f.support) := by
+  have : f.support = (U \ {x | f x = (0 : 𝕜 → ℤ) x}) := by
+    ext x
+    simp only [Function.mem_support, ne_eq, Pi.zero_apply, mem_diff, mem_setOf_eq, iff_and_self]
+    exact (h ·)
+  rw [EventuallyEq, Filter.Eventually, codiscreteWithin_iff_locallyFiniteComplementWithin, this]
+
+namespace Divisor
+
+/-- A divisor can be coerced into a function 𝕜 → ℤ -/
+instance (U : Set 𝕜) : CoeFun (Divisor U) (fun _ ↦ 𝕜 → ℤ) where
+  coe := Divisor.toFun
+
+attribute [coe] Divisor.toFun
```

Comment:
I think you get these for free after you set up the `FunLike` instance in the next declaration.

Reply:
I was able to delete the attribute without consequences. Deleting the instance led to massive errors `typeclass instance is stuck` all over the file.

@jcommelin Please advise.

Reply:
I finally had some time to look into this. Here is a diff that sets up the basics so that `.toFun` should never have to appear in the source code or the goal view after the first 20 lines or so.
https://github.com/leanprover-community/mathlib4/compare/kebekus/divisor...jmc-divisor?expand=1

It also includes a few other minor adaptations which I should have made in separate comments...

Reply:
@jcommelin Thanks for taking the time! Honestly, I do not really understand these changes, but I merge on the general assumption that you know what you are doing. Thanks again!